aws/terraform/metaflow/iam-custom-role.tf

```diff
@@ -0,0 +1,317 @@
+data "aws_iam_policy_document" "metaflow_user_role_assume_role" {
+  statement {
+    actions = [
+      "sts:AssumeRole"
+    ]
+
+    effect = "Allow"
+
+    principals {
+      identifiers = [
+        module.metaflow.metadata_svc_ecs_task_role_arn
+      ]
+      type = "AWS"
+    }
+  }
+}
+
+resource "aws_iam_role" "metaflow_user_role" {
+  count = var.custom_role ? 1 : 0
+  name  = local.metaflow_user_role_name
+  # Read more about ECS' `task_role` and `execution_role` here https://stackoverflow.com/a/49947471
+  assume_role_policy = data.aws_iam_policy_document.metaflow_user_role_assume_role.json
+
+  tags = module.common_vars.tags
+}
+
+data "aws_iam_policy_document" "metaflow_policy" {
+  statement {
+    effect = "Allow"
+
+    actions = [
+      "cloudformation:DescribeStacks",
+      "cloudformation:*Stack",
+      "cloudformation:*ChangeSet"
+    ]
+
+    resources = [
+      "arn:${var.iam_partition}:cloudformation:${local.aws_region}:${local.aws_account_id}:stack/${local.resource_prefix}*${local.resource_suffix}"
+    ]
+  }
+
+  statement {
+    actions = [
+      "s3:*Object"
+    ]
+
+    effect = "Allow"
+
+    resources = [
+      "${module.metaflow.metaflow_s3_bucket_arn}/*"
+    ]
+  }
+
+  statement {
+    effect = "Allow"
+
+    actions = [
+      "sagemaker:DescribeNotebook*",
+      "sagemaker:StartNotebookInstance",
+      "sagemaker:StopNotebookInstance",
+      "sagemaker:UpdateNotebookInstance",
+      "sagemaker:CreatePresignedNotebookInstanceUrl",
+    ]
+
+    resources = [
+      "arn:${var.iam_partition}:sagemaker:${local.aws_region}:${local.aws_account_id}:notebook-instance/${local.resource_prefix}*${local.resource_suffix}",
+      "arn:${var.iam_partition}:sagemaker:${local.aws_region}:${local.aws_account_id}:notebook-instance-lifecycle-config/basic*"
+    ]
+  }
+
+  statement {
+    effect = "Allow"
+
+    actions = [
+      "iam:PassRole",
+    ]
+
+    resources = [
+      "arn:${var.iam_partition}:iam:${local.aws_region}:${local.aws_account_id}:role/${local.resource_prefix}*${local.resource_suffix}"
```

Comment:
I've only gotten a chance to test this in Gov Cloud, and not in its entirety due to another issue I'll highlight elsewhere, but this wildcard structure seems to confuse Terraform pretty dramatically.  I was able to get it to at least apply using the following syntax: `"arn:${var.iam_partition}:iam:${local.aws_region}:${local.aws_account_id}:role/${local.resource_prefix}-*-${local.resource_suffix}"`.  I'm uncertain if the policies will work, though, because I'm stalled at the same KMS S3 issue we saw with the Batch user.  I assume we'll just need to apply the same permission setup here.

Reply:
I've created a shorted region name (4 character) mapping dict to hopefully resolve this